plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1219 @@
+        can_be_inf       `bool`  [DEFAULT `True`] values can be :data:`~numpy.inf`
+        can_be_nan       `bool`  [DEFAULT `True`] values can be :data:`~numpy.nan`
+        none_shall_pass  `bool`  [DEFAULT `False`] values can be a python `None`
+        ================ ======= ================================================
+
+    Notes
+    -----
+    * Checking of function arguments `*args` and `**kwargs` is not supported.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        from plasmapy.utils.decorators.checks import CheckValues
+
+        @CheckValues(arg1={'can_be_negative': False, 'can_be_nan': False},
+                     arg2={'can_be_inf': False},
+                     checks_on_return={'none_shall_pass': True)
+        def foo(arg1, arg2):
+            return None
+
+        # on a method
+        class Foo:
+            arg1={'can_be_negative': False, 'can_be_nan': False},
+                         arg2={'can_be_inf': False},
+                         checks_on_return={'none_shall_pass': True)
+            def bar(self, arg1, arg2):
+                return None
+    """
+    #: Default values for the possible 'check' keys.
+    # To add a new check to the class, the following needs to be done:
+    #   1. Add a key & default value to the `__check_defaults` dictionary
+    #   2. Add a corresponding if-statement to method `_check_value`
+    #
+    __check_defaults = {
+        'can_be_negative': True,
+        'can_be_complex': False,
+        'can_be_inf': True,
+        'can_be_nan': True,
+        'none_shall_pass': False,
+    }
+
+    def __init__(
+            self,
+            checks_on_return: Dict[str, bool] = None,
+            **checks: Dict[str, bool]):
+
+        super().__init__(checks_on_return=checks_on_return, **checks)
+
+    def __call__(self, f):
+        """
+        Parameters
+        ----------
+        f
+            Function to be wrapped
+
+        Returns
+        -------
+        function
+            wrapped function of `f`
+        """
+        self.f = f
+        wrapped_sign = inspect.signature(f)
+
+        @preserve_signature
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            # map args and kwargs to function parameters
+            bound_args = wrapped_sign.bind(*args, **kwargs)
+            bound_args.apply_defaults()
+
+            # get checks
+            checks = self._get_value_checks(bound_args)
+
+            # check input arguments
+            for arg_name in checks:
+                # skip check of output/return
+                if arg_name == 'checks_on_return':
+                    continue
+
+                # check argument
+                self._check_value(bound_args.arguments[arg_name],
+                                  arg_name,
+                                  checks[arg_name])
+
+            # call function
+            _return = f(**bound_args.arguments)
```

Comment:
I'm wondering if there would be an incompatibility here with the [positional only parameters](https://www.python.org/dev/peps/pep-0570/) that were implemented in Python 3.8.  I don't know if it would be worth fixing that now, but we might want to add a comment or note in the docstring if this wouldn't work with positional only parameters, or perhaps create an issue.

Reply:
Ummm....I'll need to think about this more.  I do think this would be good as a separate issue, assuming a quick alteration does not come to mind.

Reply:
I would be really suprised if the addition of positional-only params broke inspect usages... I think it's probable to be fine.